### ISSUE TITLE
msal node test app and electron 17

### ISSUE THREAD

**Issue by jeffj6123** [AzureAD/microsoft-authentication-library-for-js#4631]
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

1.70

### Wrapper Library

MSAL Node Extensions (@azure/msal-node-extensions)

### Wrapper Library Version

1.70

### Description

When taking the electron sample and updating electron to 17.2.0 I am getting the follow error when logging in and trying to open emails.

### Error Message

E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp>npm start

> electron-test-app@1.0.0 start E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp
> npm run build && electron dist/app.js

> electron-test-app@1.0.0 build E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp
> tsc

[Fri, 25 Mar 2022 21:34:43 GMT] : @azure/msal-node@1.7.0 : Info - getTokenCache called
[Fri, 25 Mar 2022 21:34:48 GMT] : @azure/msal-node@1.7.0 : Info - getAuthCodeUrl called
[Fri, 25 Mar 2022 21:34:48 GMT] : @azure/msal-node@1.7.0 : Verbose - initializeRequestScopes called
[Fri, 25 Mar 2022 21:34:48 GMT] : [84fb7100-2d07-4ac1-a864-1f1d620e11a3] : @azure/msal-node@1.7.0 : Verbose - buildOauthClientConfiguration called
[Fri, 25 Mar 2022 21:34:48 GMT] : [84fb7100-2d07-4ac1-a864-1f1d620e11a3] : @azure/msal-node@1.7.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/08416161-211e-4d03-a6f5-4e12360d74c3
[Fri, 25 Mar 2022 21:34:48 GMT] : [84fb7100-2d07-4ac1-a864-1f1d620e11a3] : @azure/msal-node@1.7.0 : Verbose - createAuthority called
[Fri, 25 Mar 2022 21:34:49 GMT] : [84fb7100-2d07-4ac1-a864-1f1d620e11a3] : @azure/msal-node@1.7.0 : Verbose - Auth code client created
[Fri, 25 Mar 2022 21:34:50 GMT] : @azure/msal-node@1.7.0 : Info - acquireTokenByCode called
[Fri, 25 Mar 2022 21:34:50 GMT] : @azure/msal-node@1.7.0 : Verbose - initializeRequestScopes called
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-node@1.7.0 : Verbose - buildOauthClientConfiguration called
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-node@1.7.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/08416161-211e-4d03-a6f5-4e12360d74c3
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-node@1.7.0 : Verbose - createAuthority called
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-node@1.7.0 : Verbose - Auth code client created
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-common@6.1.0 : Info - in acquireToken call
(node:163884) electron: Failed to load URL: msal://redirect/?code=<TOKEN_REDACTED>&session_state=165060c1-98e2-488d-b31a-40d0501c4e72 with error: ERR_FILE_NOT_FOUND
(Use `electron --trace-warnings ...` to show where the warning was created)
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-common@6.1.0 : Verbose - Persistence enabled, calling beforeCacheAccess
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-common@6.1.0 : Verbose - Persistence enabled, calling afterCacheAccess
[Fri, 25 Mar 2022 21:34:55 GMT] : @azure/msal-node@1.7.0 : Verbose - initializeRequestScopes called
[Fri, 25 Mar 2022 21:34:55 GMT] : [c0d494d7-a0f5-4d8e-944b-4bd117d97410] : @azure/msal-node@1.7.0 : Verbose - buildOauthClientConfiguration called
[Fri, 25 Mar 2022 21:34:55 GMT] : [c0d494d7-a0f5-4d8e-944b-4bd117d97410] : @azure/msal-node@1.7.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/08416161-211e-4d03-a6f5-4e12360d74c3
[Fri, 25 Mar 2022 21:34:55 GMT] : [c0d494d7-a0f5-4d8e-944b-4bd117d97410] : @azure/msal-node@1.7.0 : Verbose - createAuthority called
[Fri, 25 Mar 2022 21:34:55 GMT] : [c0d494d7-a0f5-4d8e-944b-4bd117d97410] : @azure/msal-node@1.7.0 : Verbose - Silent flow client created
Request made at: Fri Mar 25 2022 14:34:55 GMT-0700 (Pacific Daylight Time)
(node:163884) UnhandledPromiseRejectionWarning: Error: Request failed with status code 401
    at createError (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\core\createError.js:16:15)
    at settle (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\core\settle.js:17:12)
    at IncomingMessage.handleStreamEnd (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\adapters\http.js:269:11)
    at IncomingMessage.emit (node:events:402:35)
    at endReadableNT (node:internal/streams/readable:1343:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
(node:163884) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)

### MSAL Configuration

```javascript
{
    "authOptions":
        {
            "clientId": "904df931-49c6-404f-ad6c-4fd84a25d1bb",
            "authority": "https://login.microsoftonline.com/08416161-211e-4d03-a6f5-4e12360d74c3"
        },
    "request":
    {
        "authCodeUrlParameters": {
            "scopes": ["user.read"],
            "redirectUri": "msal://redirect"
        },
        "authCodeRequest": {
            "redirectUri": "msal://redirect",
            "scopes": ["User.Read"]
        }
    },
    "resourceApi":
    {
        "endpoint": "https://graph.microsoft-ppe.com/v1.0/me"
    },
    "fileProtocol":
    {
        "name": "msal"
    }
}
```

### Relevant Code Snippets

```javascript
updating the test msal node app to use electron 17.2.0. only two files were touched.

in createMainWindow in Main.ts I added
contextIsolation: false to the webpreferences.

in package.json I changed electron to
    "electron": "^17.2.0",
```

### Reproduction Steps

Clone the msal repo.

change electron to 17.2.0 in package.json

change contextIsolation to false in CreateMainWindow from the file Main.ts(electron set this default to true after v10)

npm start

login

attempt to open email

### Expected Behavior

error is not thrown and access is allowed.

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

None (Server)

### Source

Internal (Microsoft)

**Comment by sameerag** [comment]
@salman90 Can you please check this?

**Comment by salman90** [comment]
@hectormmg

**Comment by jeffj6123** [comment]
@salman90  @hectormmg Is this version not supported?

**Comment by hectormmg** [comment]
@jeffj6123 although Electron technically isn't officially supported by MSAL Node yet, we expect MSAL Node still works for the newer versions of Electron. In this case the sample is just outdated, I'm working on an update and will update this issue when we have an up-to-date working version. Thanks!

**Comment by hectormmg** [comment]
@jeffj6123 

> Request made at: Fri Mar 25 2022 14:34:55 GMT-0700 (Pacific Daylight Time)
(node:163884) UnhandledPromiseRejectionWarning: Error: Request failed with status code 401
at createError (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\core\createError.js:16:15)
at settle (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\core\settle.js:17:12)
at IncomingMessage.handleStreamEnd (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\adapters\http.js:269:11)
at IncomingMessage.emit (node:events:402:35)
at endReadableNT (node:internal/streams/readable:1343:12)
at processTicksAndRejections (node:internal/process/task_queues:83:21)

I can see from your last error message that the problem is actually not an MSAL issue, but you're getting a 401 (Unauthorized) error when you make the graph request. Could you check if the token is being added to the request correctly? We're still updating the sample but if your issue is related to user permissions, you may need to sort that out.

**Comment by jeffj6123** [comment]
@hectormmg can this part be ignored then?

[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-common@6.1.0 : Info - in acquireToken call
_(node:163884) electron: Failed to load URL: msal://redirect/?code=<TOKEN_REDACTED>&session_state=165060c1-98e2-488d-b31a-40d0501c4e72 with error: ERR_FILE_NOT_FOUND_
(Use electron --trace-warnings ... to show where the warning was created)

**Comment by hectormmg** [comment]
@jeffj6123 I believe so. The error comes from the auth window trying to load the URL when it is changed after the redirect. This isn't a URL meant to be navigated to, we just need to extract the auth code from it through the custom file protocol's listener. If you can confirm that the auth code is being extracted by the Custom file protocol listener and then that code is being successfully exchanged for an access token, you can ignore this error.

I'm looking at ways for preventing the auth window from attempting to navigate (or load the file in this case) so this error isn't shown on the console.

Did you get a chance to confirm there's no permission issue based on the 401?

**Comment by jeffj6123** [comment]
hey, @hectormmg the 401 error I believe came from me not setting a different url than  "https://graph.microsoft-ppe.com/v1.0/me" but yes I do see a token get generated.

**Comment by jasonnutter** [comment]
@jeffj6123 I believe this will be fixed by #4684, @hectormmg please confirm.

**Comment by hectormmg** [comment]
@jasonnutter there's two separate problems related to this issue, here's a summary:

1. The 401 when calling the /messages graph endpoint

It turns out that the /messages endpoint is disabled for @microsoft accounts, which is why this error comes up in any/all of our samples that use /messages (Mail.Read).

2. The "could not load URL" error that happens after the auth code redirect is actually the result of using a **file** protocol to read the auth code. The protocol's last default action is to attempt to load the file from that address on disk, which doesn't exist, causing the warning even if the auth code was successfully obtained. Switching to the **string** protocol fixes this.

So yes, #4684 will resolve this issue.
